packages/babel-core/src/config/config-chain.ts

```diff
@@ -443,34 +448,43 @@ function buildFileLogger(
   });
 }
 
-function buildRootDescriptors({ dirname, options }, alias, descriptors) {
+function buildRootDescriptors(
+  { dirname, options }: Partial<ValidatedFile>,
+  alias: string,
+  descriptors: Function,
+) {
   return descriptors(dirname, options, alias);
 }
 
-function buildProgrammaticLogger(_, context, baseLogger: ConfigPrinter | void) {
+function buildProgrammaticLogger(
+  _: unknown,
+  context: ConfigContext,
+  baseLogger: ConfigPrinter | void,
+) {
   if (!baseLogger) {
     return () => {};
   }
   return baseLogger.configure(context.showConfig, ChainFormatter.Programmatic, {
+    // @ts-expect-error caller maybe void
     callerName: context.caller?.name,
   });
 }
 
 function buildEnvDescriptors(
-  { dirname, options },
-  alias,
-  descriptors,
-  envName,
+  { dirname, options }: Partial<ValidatedFile>,
+  alias: string,
+  descriptors: Function,
```

Comment:
```suggestion
  descriptors: (dirname: string, options: ValidatedOptions, alias: string) => OptionsAndDescriptors,
```

See also `createUncachedDescriptors` in `packages/babel-core/src/config/config-descriptors.ts`